### PR TITLE
feat: wrap widgets into a component which listens to content height changes

### PR DESCRIPTION
**Pull request by ankurrsinghal**
## Description

- Problem
We know that some of our widgets are dynamic, take for example CheckboxGroup widget, we can add and remove options from the property pane or hook its data to an API. How can we know how much this widget wants to grow in height?

- Solution
What if we wrap each widget in a simple div whose height is auto, so now this div will always have height equal to the widget's height, whatever widget wants to grow or shrink the height of this div will be equal to that. What about the changes in height? We used `ResizeObserver` on this simple div which keep sending us updates whenever a widget wants to change it's height. And then we send these updates to our `updateDynamicHeight` store actions from the BaseWidget.

- Features
- This component is very simple, accepts 3 props, maxDynamicHeight, isAutoWithLimits and onHeightUpdate.
- maxDynamic this is used only when isAutoWithLimits is enabled, and adds 2 containers to the div while the inner one sends the height updates and the outer one checks if the max is less than the expected height of the widget, in that case it scrolls.
- onHeightUpdate is a callback function sent from BaseWidget, which actually calls the `updateDynamicHeight` action.

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Written a complimentary DynamicHeightContainer.test.tsx which includes some jest test suite to test the functionality of the DynamicHeightContainer.tsx.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/components/dynamicHeight/DynamicHeightContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      }
+    };
+  }, [observer]);
+
+  if (isAutoHeightWithLimits) {
+    const expectedHeightInRows = Math.ceil(
+      expectedHeight / GridDefaults.DEFAULT_GRID_ROW_HEIGHT,
+    );
+
+    console.log("expectedHeightInRows", expectedHeightInRows);
```
Comment (riodeuno) [inline app/client/src/components/dynamicHeight/DynamicHeightContainer.tsx:55]: Please remove a missed out `console.log`

Comment (ankurrsinghal) [reply app/client/src/components/dynamicHeight/DynamicHeightContainer.tsx]: Done

**Review by riodeuno — COMMENTED** [review]
Please checkout the updates I've made in this component in this PR here: https://github.com/appsmithorg/appsmith/pull/17686

The reason for this, is that when we get a new min or max dynamic height, we need to make sure we need to run the updateDynamicHeight to make sure that the changes are reflected in the UI.

**Comment by riodeuno** [comment]
@ankurrsinghal I really like the PR's description. It explains what we've done here.
The checklist from the PR template is missing. Can we use the template and add all the current contents into the description section of the template?

**Comment by ankurrsinghal** [comment]
> @ankurrsinghal I really like the PR's description. It explains what we've done here. The checklist from the PR template is missing. Can we use the template and add all the current contents into the description section of the template?

Done

**Comment by ankurrsinghal** [comment]
> Please checkout the updates I've made in this component in this PR here: #17686
> 
> The reason for this, is that when we get a new min or max dynamic height, we need to make sure we need to run the updateDynamicHeight to make sure that the changes are reflected in the UI.

Did the corresponding updates!

**Review by rahulramesha — APPROVED** [review]

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat-create-run-meta-before-test-starts-compare-lt-st-for-prev-runs

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat-create-run-meta-before-test-starts-compare-lt-st-for-prev-runs

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat-create-run-meta-before-test-starts-compare-lt-st-for-prev-runs

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat-create-run-meta-before-test-starts-compare-lt-st-for-prev-runs

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat-create-run-meta-before-test-starts-compare-lt-st-for-prev-runs

**Comment by Rhitottam** [comment]
/perf-test ref=rhito/feat-create-run-meta-before-test-starts-compare-lt-st-for-prev-runs

File: app/client/src/components/dynamicHeight/DynamicHeightContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    if (ref.current) {
+      observer.current.observe(ref.current);
+    }
+
+    return () => {
+      if (ref.current) {
+        observer.current.unobserve(ref.current);
+      }
+    };
+  }, [observer]);
```
Comment (marks0351) [inline app/client/src/components/dynamicHeight/DynamicHeightContainer.tsx:50]: u can remove observer it will have no effect on the code.

**Review by marks0351 — CHANGES_REQUESTED** [review]

File: app/client/src/components/dynamicHeight/DynamicHeightContainer.test.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
+import React from "react";
+import "@testing-library/jest-dom";
+import { render } from "@testing-library/react";
+import DynamicHeightContainer from "./DynamicHeightContainer";
+import "jest-styled-components";
+import renderer from "react-test-renderer";
+
```
Comment (riodeuno) [inline app/client/src/components/dynamicHeight/DynamicHeightContainer.test.tsx:7]: @ankurrsinghal Let's rename the folder from `dynamicHeight` to `autoHeight`, and `DynamicHeightContainer` to `AutoHeightContainer`

**Review by riodeuno — CHANGES_REQUESTED** [review]
Please change the names of the files. A new push to this branch should help with the vercel run as well. @marks0351 also has a comment, please look into that.

File: app/client/src/components/dynamicHeight/DynamicHeightContainer.tsx (hunk 2 above)
Comment (ankurrsinghal) [reply app/client/src/components/dynamicHeight/DynamicHeightContainer.tsx]: Done

File: app/client/src/components/dynamicHeight/DynamicHeightContainer.test.tsx (hunk 3 above)
Comment (ankurrsinghal) [reply app/client/src/components/dynamicHeight/DynamicHeightContainer.test.tsx]: Done

**Review by riodeuno — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Pull request merged**